src/base/stun.ts

```diff
@@ -41,175 +42,322 @@ export const DEFAULT_PARALLEL_STUN_CALLS = 4
+    responses.push(...(await performSTUNRequests(usableMultiaddrs, socket, STUN_TIMEOUT)))
 
-      if (results.length == 1) {
-        resolve(results[0])
+    filteredResults = getUsableResults(responses, runningLocally)
+  }
+
+  const interpreted = intepreteResults(filteredResults)
+
+  if (interpreted.ambiguous) {
+    return undefined
+  } else {
+    return interpreted.publicAddress
+  }
+}
+
+async function performSTUNRequests(
+  multiAddrs: Multiaddr[],
+  socket: Socket,
+  timeout = STUN_TIMEOUT
+): Promise<Request[]> {
+  let requests = generateRequests(multiAddrs)
+  let results = decodeIncomingSTUNResponses(requests, socket, timeout)
+  sendStunRequests(requests, socket)
+
+  return (await results).filter((request: Request) => request.response)
+}
+
+/**
+ * Send requests to given STUN servers
+ * @param usableMultiaddrs multiaddrs to use for STUN requests
+ * @param tIds transaction IDs to use, necessary to link requests and responses
+ * @param socket the socket to send the STUN requests
+ * @returns usable transaction IDs and the corresponding multiaddrs
+ */
+function sendStunRequests(addrs: Request[], socket: Socket): void {
+  for (const addr of addrs) {
+    if (![CODE_IP4, CODE_IP6, CODE_DNS4, CODE_DNS6].includes(addr.multiaddr.tuples()[0][0])) {
+      error(`Cannot contact STUN server ${addr.multiaddr.toString()} due invalid address.`)
+      continue
+    }
+
+    let nodeAddress: ReturnType<Multiaddr['nodeAddress']>
+    try {
+      nodeAddress = addr.multiaddr.nodeAddress()
+    } catch (err) {
+      error(err)
+      continue
+    }
+
+    const res = stun.createBindingRequest(addr.tId).setFingerprintAttribute()
+
+    verbose(`STUN request sent to ${nodeAddress.address}:${nodeAddress.port}`)
+    socket.send(res.toBuffer(), nodeAddress.port, nodeAddress.address, (err: any) => err && error(err.message))
+  }
+}
+
+function decodeIncomingSTUNResponses(addrs: Request[], socket: Socket, ms: number = STUN_TIMEOUT) {
+  return new Promise<Request[]>((resolve) => {
+    let responsesReceived = 0
+
+    let done: () => void
+    let listener: (msg: Buffer) => void
+    let finished = false
+
+    const timeout = setTimeout(() => {
+      log(`STUN timeout. None of the selected STUN servers replied.`)
+      done()
+    }, ms)
+
+    done = () => {
+      if (finished) {
         return
       }
+      finished = true
 
-      for (const result of results) {
-        if (result.port != results[0].port) {
-          return resolve(undefined)
-        }
-      }
+      socket.removeListener('message', listener)
 
-      resolve(results[0])
+      clearTimeout(timeout)
+
+      resolve(addrs)
     }
 
-    const msgHandler = (msg: Buffer) => {
+    listener = (msg: Buffer) => {
       const res = stun.createBlank()
 
       // Overwrite console.log because 'webrtc-stun' package
       // pollutes console output
       const consoleBackup = console.log
```

Comment:
Probably not the smartest idea, but if absolutely necessary please move this code to `setup` and `tearDown` of the tests to avoid test code pollution

Reply:
this is not part of a test ... this is normal STUN behavior code